src/charm.py

```diff
@@ -5,52 +5,108 @@
 """Flask Charm service."""
 
 import logging
+import shlex
 import typing
 
 from charms.traefik_k8s.v1.ingress import IngressPerAppRequirer
 from ops.charm import CharmBase, ConfigChangedEvent
 from ops.main import main
-from ops.model import ActiveStatus
+from ops.model import ActiveStatus, BlockedStatus, Container, StatusBase
+
+from charm_state import CharmState
+from consts import FLASK_APP_PORT, FLASK_CONTAINER_NAME, FLASK_SERVICE_NAME
+from exceptions import CharmConfigInvalidError
+from flask_app import FlaskApp
+from webserver import GunicornWebserver
 
 logger = logging.getLogger(__name__)
 
 
 class FlaskCharm(CharmBase):
     """Flask Charm service."""
 
-    _FLASK_APP_PORT = 8000
-
     def __init__(self, *args: typing.Any) -> None:
         """Initialize the instance.
 
         Args:
             args: passthrough to CharmBase.
         """
         super().__init__(*args)
-        self.framework.observe(self.on.config_changed, self.config_service)
+        try:
+            self._charm_state = CharmState.from_charm(charm=self)
+        except CharmConfigInvalidError as exc:
+            self._update_status(BlockedStatus(exc.msg))
+            return
```

Comment:
I'm not quite sure if the state should raise error during init, would it be better if we had validation in the event handler?

Reply:
The reason I raise an error during the initialization is that I don't want the state to contain invalid data.